File: coding_skill_cpp/20_lazy_initialization.cpp
/* C++ Lazy Init - optional */
#include <iostream>
#include <optional>
using namespace std;

class LazyResource {
    optional<int> data;
public:
    int& get() {
        if (!data) {
            cout << "[Lazy] 첫 접근! 초기화" << endl;
            data = 100;
        }
        return *data;
    }
};

int main() {
    cout << "=== C++ Lazy Initialization ===" << endl;
    LazyResource res;
    cout << "1st access: " << res.get() << endl;
    cout << "2nd access: " << res.get() << endl;
    return 0;
}

File: coding_skill_py/20_lazy_initialization.py
"""Python Lazy Init - property"""
from functools import cached_property

class LazyResource:
    @cached_property
    def data(self):
        print("[Lazy] 첫 접근! 초기화")
        return 100

def main():
    print("=== Python Lazy Initialization ===")
    res = LazyResource()
    print(f"1st access: {res.data}")
    print(f"2nd access: {res.data}")

if __name__ == "__main__":
    main()
